kvnemesis: address fault mode todos

miraradeva opened the pull request:
This commit addresses TODOs left after introducing network partitions and node restarts to kvnemesis:

- All fault mode tests run with 4 nodes to ensure that liveness tests can always maintain a majority quorum of replicas on nodes 1 and 2. If the tests run with 5 nodes, we need to take extra care to ensure 3 replicas are healthy for all RF=5 ranges (e.g. the systems ranges).

- Splits are now allowed in both partition and restart liveness modes. Previously, if r3 was unavailable, the range ID allocator would get stuck retrying the increment operation that generates a new range ID. This commit ensures all systems ranges are available in liveness mode.

- Lease transafers are now allowed in partition liveness mode. It's not clear what changed but the failure doesn't repro anymore. Will investigate more if it fails in CI.

Part of: #64828
Part of: #114814

Release note: None

miraradeva committed:
kvnemesis: address fault mode todos

This commit addresses TODOs left after introducing network partitions
and node restarts to kvnemesis:

- All fault mode tests run with 4 nodes to ensure that liveness tests
  can always maintain a majority quorum of replicas on nodes 1 and 2.
  If the tests run with 5 nodes, we need to take extra care to ensure 3
  replicas are healthy for all RF=5 ranges (e.g. the systems ranges).

- Splits are now allowed in both partition and restart liveness modes.
  Previously, if r3 was unavailable, the range ID allocator would get
  stuck retrying the increment operation that generates a new range ID.
  This commit ensures all systems ranges are available in liveness
  mode. Moreover, for the restart liveness variant, some error handling
  tweaks were needed to make sure the split is retried if it races with
  a node shutdown.

- A few operations are re-enabled across the two partition test
  variants: lease transafers and delete range. It's not clear what
  changed but these don't cause the tests to fail anymore (or at least
  it's hard to repro). Potentially, the theory for why they could fail
  was not quite right. Will investigate more if they fail in CI.

Part of: #64828
Part of: #114814

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/156088)
<!-- Reviewable:end -->

miraradeva committed:
kvserver: return NodeUnavailable error during range ID allocation

During a split, the request to generate a new range ID for the RHS can
race with shutting down the node where the underlying increment is
evaluated. Previously, in this case, the range ID allocator would
return a custom error (`could not allocate ID; system is draining`),
which is not retried by the `DistSender`.

This commit replaces this error with a `NodeUnavailableError`, which is
how this type of race is handled in other places as well.

Informs: #64828

Release note: None

miraradeva committed:
kvnemesis: fix CPut read validation in weak isolation levels

Previously, kvnemesis would attempt to validate the timestamp associated
with a CPut's read regardless of the isolation level. Since the read
component of CPut is non-locking, this validation is not correct
because the CPut's read is not refreshed in weak-isolation
transactions.

This commit treats the read of CPut similarly to other non-locking
reads; they are validated only for serializable transctions.

Part of: #64810

Release note: None

stevendanna reviewed (approved):
Awesome, nice that we have less special configuration now.

> @stevendanna reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/156088#-:-OcarkK32kft8iPWyLjV:b-an2foe)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
